base/src/testFixtures/kotlin/proguard/testutils/PartialEvaluatorUtil.kt

```diff
@@ -26,9 +26,9 @@ class PartialEvaluatorUtil {
             val method = clazz.findMethod(methodName, methodDescriptor) as ProgramMethod
 
             val codeAttribute =
-                method.attributes.find { it.getAttributeName(clazz) == Attribute.CODE } as CodeAttribute
+                method.attributes.slice(0 until method.u2attributesCount).find { it.getAttributeName(clazz) == Attribute.CODE } as CodeAttribute
```

Comment:
Visitors to the rescue ;) It should not be necessary to iterate over the array fields directly here.

```
        fun evaluate(
            className: String,
            methodName: String,
            methodDescriptor: String,
            classPool: ClassPool,
            partialEvaluator: PartialEvaluator
        ): Pair<ArrayList<Pair<Int, Instruction>>, HashMap<String, Int>> {
            val clazz = classPool.getClass(className) as ProgramClass
            val method = clazz.findMethod(methodName, methodDescriptor) as ProgramMethod

            val instructions = ArrayList<Pair<Int, Instruction>>()
            val variableTable = HashMap<String, Int>()
            method.attributesAccept(
                clazz,
                AttributeNameFilter(
                    Attribute.CODE,
                    MultiAttributeVisitor(
                        // Execute the partial evaluator.
                        partialEvaluator,
                        // Collect instructions.
                        AllInstructionVisitor(object : InstructionVisitor {
                            override fun visitAnyInstruction(
                                clazz: Clazz,
                                method: Method,
                                codeAttribute: CodeAttribute,
                                offset: Int,
                                instruction: Instruction
                            ) {
                                instructions.add(Pair(offset, instruction))
                            }
                        }),
                        // Collect local variables.
                        AllAttributeVisitor(object : AttributeVisitor {
                            override fun visitAnyAttribute(clazz: Clazz, attribute: Attribute) {
                                // Ignore other attributes.
                            }

                            override fun visitLocalVariableTableAttribute(
                                clazz: Clazz,
                                method: Method,
                                codeAttribute: CodeAttribute,
                                localVariableTableAttribute: LocalVariableTableAttribute
                            ) {
                                localVariableTableAttribute.localVariablesAccept(
                                    clazz,
                                    method,
                                    codeAttribute
                                ) { _, _, _, localVar ->
                                    variableTable[localVar.getName(clazz)] = localVar.u2index
                                }
                            }
                        })
                    )
                )
            )

            return Pair(instructions, variableTable)
        }
```

Reply:
Ah that makes sense, that's exactly what they are made for.